sys/net/link_layer/ieee802154/submac.c

```diff
@@ -0,0 +1,384 @@
+
+    if (ieee802154_radio_has_frame_retrans(dev) ||
+        ieee802154_radio_has_irq_ack_timeout(dev) || !submac->wait_for_ack) {
+        _tx_end(submac, info->status, info);
+    }
+    else {
+        ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_WAIT_FOR_ACK);
+
+        /* Handle ACK reception */
+        ieee802154_submac_ack_timer_set(submac, ACK_TIMEOUT);
+    }
+}
+
+static void _handle_tx_medium_busy(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    if (ieee802154_radio_has_frame_retrans(dev) ||
+        ieee802154_radio_has_auto_csma(dev)) {
+        ieee802154_radio_request_set_trx_state(dev, IEEE802154_TRX_STATE_RX_ON);
+        _tx_end(submac, TX_STATUS_MEDIUM_BUSY, NULL);
+    }
+    else {
+        /* CCA failed. Continue with the CSMA-CA algorithm */
+        _perform_csma_ca(submac);
+    }
+}
+
+static void _handle_tx_no_ack(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    if (ieee802154_radio_has_frame_retrans(dev)) {
+        ieee802154_radio_request_set_trx_state(dev, IEEE802154_TRX_STATE_RX_ON);
+        submac->wait_for_ack = false;
+        _tx_end(submac, TX_STATUS_NO_ACK, NULL);
+    }
+    else {
+        /* Perform retransmissions */
+        _perform_retrans(submac);
+    }
+}
+
+void ieee802154_submac_tx_done_cb(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+    ieee802154_tx_info_t info;
+
+    ieee802154_radio_confirm_transmit(dev, &info);
+
+    switch (info.status) {
+    case TX_STATUS_MEDIUM_BUSY:
+        _handle_tx_medium_busy(submac);
+        break;
+    case TX_STATUS_NO_ACK:
+        _handle_tx_no_ack(submac);
+        break;
+    case TX_STATUS_SUCCESS:
+    case TX_STATUS_FRAME_PENDING:
+        _handle_tx_success(submac, &info);
+        break;
+    default:
+        assert(false);
+        break;
+    }
+}
+
+int ieee802154_send(ieee802154_submac_t *submac, const iolist_t *iolist)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    uint8_t *buf = iolist->iol_base;
+    bool cnf = buf[0] & IEEE802154_FCF_ACK_REQ;
+
+    if (submac->state == IEEE802154_STATE_OFF) {
+        return -ENETDOWN;
+    }
+
+    if (submac->tx ||
+        ieee802154_radio_request_set_trx_state(dev,
+                                               IEEE802154_TRX_STATE_TX_ON) <
+        0) {
+        return -EBUSY;
+    }
+
+    submac->tx = true;
+
+    ieee802154_radio_write(dev, iolist);
+    while (ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {}
+
+    submac->wait_for_ack = cnf;
+    submac->retrans = 0;
+
+    ieee802154_csma_ca_transmit(submac);
+    return 0;
+}
+
+int ieee802154_submac_init(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    submac->tx = false;
+    submac->state = IEEE802154_STATE_LISTEN;
+
+    ieee802154_radio_request_on(dev);
+
+    /* generate EUI-64 and short address */
+    luid_get_eui64(&submac->ext_addr);
+    luid_get_short(&submac->short_addr);
```

Comment:
sure, I can use these functions.

>  I can add a PR for cc2538_rf and nrf802154, those are pretty trivial changes.

That would be cool.

> It only requires the driver to call netdev_register()

Does this adds a dependency to the `netdev` interface? Or it's just about "network devices"?

Reply:
how to proceed with this?

Reply:
It only needs the `.type` and `.index` members of the `struct netdev`

Reply:
#14994, #14995

Reply:
thanks!

Reply:
In order to use this feature (or any that generates these addresses), I think the best thing to do is to modify the `ieee802154_submac_init` function to receive these addresses (instead of hardcoding the dependency to `luid`).
Then, we can simply use the `netdev_register` during `netdev_submac->driver->init()`

Reply:
Both PRs have been merged.